lib/bank_account.ex

```diff
@@ -8,31 +10,78 @@ defmodule BankAccount do
   """
   @opaque account :: pid
 
+  @impl true
+  def init(:ok) do
+    {:ok, %{balance: 0, status: :account_open}}
+  end
+
+  def start_link do
+    GenServer.start_link(__MODULE__, :ok)
+  end
+
+  @impl true
+  # balance inquiry when status: :account_closed
+  def handle_call(:balance, _from, %{status: :account_closed} = state) do
+    {:reply, {:error, :account_closed}, state}
+  end
+
+  @impl true
+  # balance inquiry
+  def handle_call(:balance, _from, state) do
+    {:reply, Map.get(state, :balance), state}
+  end
+
+  @impl true
+  # close bank/update status
+  def handle_call({:status, :account_closed}, _from, state) do
+    {:reply, :ok, Map.put(state, :status, :account_closed)}
+  end
+
+  @impl true
+  # balance update when status: account_closed
+  def handle_call({:balance, _amount}, _from, %{status: :account_closed} = state) do
+    {:reply, {:error, :account_closed}, state}
+  end
+
+  @impl true
+  # balance update
+  def handle_call({:balance, amount}, _from, state) do
+    balance = Map.get(state, :balance)
+    {:reply, :ok, Map.put(state, :balance, balance + amount)}
+  end
+
   @doc """
   Open the bank. Makes the account available.
   """
   @spec open_bank() :: account
+  # TODO: handle errors?
+  # @spec open_bank() :: {:ok, account} | {:error, term()}
   def open_bank() do
+    {:ok, account} = BankAccount.start_link()
+    account
```

Comment:
```suggestion
  # @spec open_bank!() :: account
  def open_bank!() do
    {:ok, account} = BankAccount.start_link()
    account
```
A function that raises on error (such as this will do if `start_link` fails), it should have a bang in the function name. Alternatively, you can swap to returning an :ok/:error tuple:
```suggestion
  # @spec open_bank() :: {:ok, account} | {:error, term()}
  def open_bank() do
    case BankAccount.start_link() do
      {:ok, account} -> {:ok, account}
      {:error, _} -> {:error, :failed_to_open_account}
    end
```